src/main/java/io/cdap/plugin/salesforce/SalesforceBulkUtil.java

```diff
@@ -109,21 +110,58 @@ public static void closeJob(BulkConnection bulkConnection, String jobId) throws
    *
    * @param bulkConnection bulk connection instance
    * @param query a SOQL query
+   * @param enablePKChunk enable PK Chunk
    * @return an array of batches
    * @throws AsyncApiException  if there is an issue creating the job
    * @throws IOException failed to close the query
    */
-  public static BatchInfo[] runBulkQuery(BulkConnection bulkConnection, String query)
+  public static BatchInfo[] runBulkQuery(BulkConnection bulkConnection, String query, boolean enablePKChunk)
     throws AsyncApiException, IOException {
 
     SObjectDescriptor sObjectDescriptor = SObjectDescriptor.fromQuery(query);
     JobInfo job = createJob(bulkConnection, sObjectDescriptor.getName(), OperationEnum.query, null);
-
+    BatchInfo batchInfo;
     try (ByteArrayInputStream bout = new ByteArrayInputStream(query.getBytes())) {
-      bulkConnection.createBatchFromStream(job, bout);
+      batchInfo = bulkConnection.createBatchFromStream(job, bout);
     }
+    return enablePKChunk ? waitForBatchChunks(bulkConnection, job.getId(), batchInfo.getId()) :
+      bulkConnection.getBatchInfoList(job.getId()).getBatchInfo();
+  }
+
+  /** When PK Chunk is enabled, wait for state of initial batch to be NotProcessed, in this case Salesforce API will
+   * decide how many batches will be created
+   * @param bulkConnection bulk connection instance
+   * @param jobId a job id
+   * @param initialBatchId a batch id
+   * @return Array with Batches created by Salesforce API
+   *
+   * @throws AsyncApiException if there is an issue creating the job
+   */
+  private static BatchInfo[] waitForBatchChunks(BulkConnection bulkConnection, String jobId, String initialBatchId)
+    throws AsyncApiException {
+    BatchInfo initialBatchInfo = null;
+    for (int i = 0; i < GET_BATCH_RESULTS_TRIES; i++) {
+      //check if the job is aborted
+      if (bulkConnection.getJobStatus(jobId).getState() == JobStateEnum.Aborted) {
+        throw new RuntimeException(new RuntimeException(String.format("Job with Id: '%s' is aborted", jobId)));
```

Comment:
Why are we throwing a RuntimeException when the the api returns an error? Can we throw a checked exception instead?

Reply:
We changed logic here, we return zero splits and log info message.

Reply:
Is this not considered to be an error condition where we need to throw an exception?

Reply:
Job is Aborted only when user abort it from Salesforce site, so we thought this is not exception.